lib/IDE/Refactoring.cpp

```diff
@@ -2602,6 +2602,154 @@ bool RefactoringActionLocalizeString::performChange() {
+    EditConsumer.accept(SM, startLocation, "#>\n");
+  }
+}
+
+void processReturn(Type returnType, swift::SourceLoc startLocation,
+                   SourceEditConsumer &EditConsumer,
+                   SourceManager &SM) {
+    
+  if (returnType.isNull())
+    return;
+  
+  EditConsumer.accept(SM, startLocation, "///\n///  - Returns: <#");
+  EditConsumer.accept(SM, startLocation, returnType.getString());
+  EditConsumer.accept(SM, startLocation, "#>\n");
+}
+    
+void processThrows(swift::SourceLoc startLocation,
+                   SourceEditConsumer &EditConsumer,
+                   SourceManager &SM) {
+    
+  EditConsumer.accept(SM, startLocation, "///\n///  - Throws: <#Throws#>\n");
+}
+
+void generateDocCommentForFunction(FuncDecl *funcDecl,
+                                   swift::SourceLoc startLocation,
+                                   SourceEditConsumer &EditConsumer,
+                                   SourceManager &SM) {
+  
+  EditConsumer.accept(SM, startLocation, "/// <#Function Summary#>\n");
+  
+  processParameters(funcDecl->getParameterLists(), startLocation,
+                    EditConsumer, SM);
+  
+  if (funcDecl->hasThrows()) {
+    processThrows(startLocation, EditConsumer, SM);
+  }
+  
+  Type returnType = funcDecl->getReturnTypeLoc().getType();
+  processReturn(returnType, startLocation, EditConsumer, SM);
+  
+  EditConsumer.accept(SM, startLocation,"///\n");
+}
+    
+void generateDocCommentForConstructor(ConstructorDecl *constructorDecl,
+                                      swift::SourceLoc startLocation,
+                                      SourceEditConsumer &EditConsumer,
+                                      SourceManager &SM) {
+  
+  EditConsumer.accept(SM, startLocation, "/// <#Initializer Summary#>\n");
+  
+  processParameters(constructorDecl->getParameterLists(), startLocation, EditConsumer, SM);
+  
+  if (constructorDecl->hasThrows()) {
+    processThrows(startLocation, EditConsumer, SM);
+  }
+  
+  EditConsumer.accept(SM, startLocation,"///\n");
+}
+    
+void generateDocCommentForSubscript(SubscriptDecl *subscriptDecl,
+                                    swift::SourceLoc startLocation,
+                                    SourceEditConsumer &EditConsumer,
+                                    SourceManager &SM) {
+  
+  EditConsumer.accept(SM, startLocation, "/// <#Subscript Summary#>\n");
+  
+  processParameters(subscriptDecl->getIndices(), startLocation,
+                    EditConsumer, SM);
+  
+  Type returnType = subscriptDecl->getElementTypeLoc().getType();
+  processReturn(returnType, startLocation, EditConsumer, SM);
+  
+  EditConsumer.accept(SM, startLocation,"///\n");
+}
+   
+bool canGenerateDocComment(DeclKind kind) {
+  return kind == DeclKind::Func || kind == DeclKind::Constructor ||
+    kind == DeclKind::Subscript;
+}
+    
+bool RefactoringActionDocCommentBoilerplate::
+  isApplicable(ResolvedCursorInfo Tok, DiagnosticEngine &Diag) {
```

Comment:
We should only allow this refactoring on the declaration names. We should return false here when `Tok.IsRef == true`.

Reply:
ah, I should have explained more. `Tok.Kind == CursorInfoKind::ValueRef` tells us the cursor points to a name of a value declaration. However, it doesn't tell us whether the name is in the declaration, e.g. `func foo() {}`, or the name is in the reference of the declaration, e.g. `bar(foo())`. This refactoring should only be available in the former case. `Tok.IsRef` indicates whether the cursor is at the later case, so we have to check and exclude them.